Cmriboldi/add event i ds

cmriboldi opened the pull request:
As mentioned in [this issue](https://github.com/mixpanel/mixpanel-swift/issues/614) there is no way to time events independently of their name. This is a HUGE issue for any serious user of Mixpanel events because there is no way to follow the [Best Practice Naming Guidelines](https://docs.mixpanel.com/docs/data-structure/events-and-properties#best-practices) and be certain your durations are correct.

For example, if I were to have a generic System event that I want to time as long as I have one of those events I'm fine, but if I ever need to time multiple system events there is no way for me to organize them or distinguish using Mixpanel which `.track(event:)` call matches with the original `time(event:)` call. This means that durations will be completely missing or inaccurate.

The that I'm adding in this MR is to link the tracking of timed events to an EventID instead of to the event name. That means callers of the SDK can provide their own unique id's for id's of events and be guaranteed that when the duration completes for that generic event the duration is completely accurate.

I added a new function signature for these timed and tracked events that allow users to provide a String eventID to link the two events. I would have liked to create a more type safe ID so that it's not just any string but because of the way the SDK works with events I didn't want to change too much besides separating the idea of an eventName with the idea of an eventID. I also separated out the idea of a `TimedEvents` dictionary from the `InternalProperties` dictionary that existed previously, this adds some small type safety around the dictionary and adds purpose clarity in the code.

The original method signatures still exist and work just as they did before they simply call the eventID versions of these methods so that there aren't two places that the logic needs to be maintained.

cmriboldi commented on Sources/MixpanelInstance.swift:
```diff
@@ -37,6 +37,8 @@ public protocol MixpanelDelegate: AnyObject {
 
 public typealias Properties = [String: MixpanelType]
 typealias InternalProperties = [String: Any]
+typealias TimedEventID = String
+typealias TimedEvents = [TimedEventID: TimeInterval]
```
I added these two typealiases for additional type-safely and clarity. Ideally TimedEventID is it's own type that wraps a string but I didn't want to change the call site too drastically from what it already is.

cmriboldi commented on Sources/MixpanelInstance.swift:
```diff
@@ -1017,17 +1020,34 @@ extension MixpanelInstance {
      - parameter properties: properties dictionary
      */
     public func track(event: String?, properties: Properties? = nil) {
+        track(event: event, withID: nil, properties: properties)
```
The call site for all of these methods remains the same for backwards compatibility, and under the hood they just call the other method that uses the event id.

cmriboldi commented on Sources/Track.swift:
```diff
@@ -32,28 +32,30 @@ class Track {
+        if !(mixpanelInstance?.trackAutomaticEventsEnabled ?? false) && eventName.hasPrefix("$ae_") {
             return timedEvents
         }
+        let timedEventID = eventID ?? eventName
```
If no eventID is provided we fall back to using whatever the eventName is which means we default back to the original behavior.

cmriboldi commented on Sources/Track.swift:
```diff
@@ -139,32 +141,32 @@ class Track {
     }
 
-    func time(event: String?, timedEvents: InternalProperties, startTime: Double) -> InternalProperties {
+    func time(eventID: TimedEventID, timedEvents: TimedEvents, startTime: TimeInterval) -> TimedEvents {
```
I renamed the property name for clarity, changed the types for consistency/correctness and removed the optionality of the eventID because it didn't actually need to be optional.

cmriboldi commented:
@zihejia and @jaredmixpanel Is there any chance we can add this? I can go ahead and fix the merge conflict but this is a great enhancement we would love to get into your SDK so that we don't have to live off of a fork of the repo.

jaredmixpanel commented:
@cmriboldi Thanks for the PR. I think we can add it, just a few things though...

1. Fix the conflict
2. Add/update tests

- https://github.com/mixpanel/mixpanel-swift/blob/master/MixpanelDemo/MixpanelDemoTests/MixpanelDemoTests.swift#L787
- https://github.com/mixpanel/mixpanel-swift/blob/master/MixpanelDemo/MixpanelDemoTests/MixpanelOptOutTests.swift#L208
- https://github.com/mixpanel/mixpanel-swift/blob/master/MixpanelDemo/MixpanelDemoMacTests/MixpanelDemoTests.swift#L714
- https://github.com/mixpanel/mixpanel-swift/blob/master/MixpanelDemo/MixpanelDemoMacTests/MixpanelOptOutTests.swift#L214

3. Instead of labeling the new parameter `eventID`, I'd suggest to be clearer/more explicit and call it: `timedEventID` 
4. Maybe make the type [UUID](https://developer.apple.com/documentation/foundation/uuid)?

cmriboldi commented:
Awesome! thank you @jaredmixpanel. All of those are reasonable. I'll make those changes and then update this MR.

cmriboldi commented:
@jaredmixpanel Sorry it took so long but here's the fixed version of this PR. I resolved the conflicts, incorporated your suggested changes, and added tests as well. Some of the unrelated tests were stale but I fixed some of them.
